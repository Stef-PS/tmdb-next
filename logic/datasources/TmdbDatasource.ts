import axios from 'axios'
import { SearchMovieResult } from '../interfaces/SearchMovie'
import { Configuration } from '../interfaces/Configuration'
import { Movie } from '../interfaces/Movie'
import { transformMovie, transformSearchMovie } from '../utils/MovieUtils'

const LOCALE = 'en-US'
const API_URL = 'https://api.themoviedb.org/3'

export class TmdbDatasource {
  private apiKey = process.env.TMDB_API_KEY ?? 'test'
  private options = {
    timeout: 2500,
    headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' }
  }
  private config: Configuration | null = null

  resetConfiguration() {
    this.config = null
  }

  getConfiguration(): Promise<Configuration> {
    if (this.config) return Promise.resolve(this.config)
    const url = `${API_URL}/configuration?api_key=${this.apiKey}`
    return axios.get(url, this.options)
      .then((response) => {
        const {
          base_url: baseUrl,
          secure_base_url: secureBaseUrl,
          poster_sizes: posterSizes
        } = response.data.images
        this.config = { images: { baseUrl, secureBaseUrl, posterSizes } }
        return this.config
      })
      .catch(error => {
        return Promise.reject(error)
      })
  }

  async searchMovie(query: string, page = 1): Promise<SearchMovieResult> {
    const url = `${API_URL}/search/movie?api_key=${this.apiKey}&query=${query}&page=${page}&language=${LOCALE}&include_adult=false`
    return axios.get(url, this.options)
      .then(({ data }) => {
        const { results, page, total_results: totalResults, total_pages: totalPages } = data
        return { page, results: results.map((movie: any) => transformSearchMovie(movie)), totalResults, totalPages }
      })
      .catch(error => {
        return Promise.reject(error)
      })
  }

  async getMovie(id: number): Promise<Movie> {
    const url = `${API_URL}/movie/${id}?api_key=${this.apiKey}&language=${LOCALE}`
    return axios.get(url, this.options)
      .then(({ data }) => {
        return transformMovie(data)
      }).catch(error => {
        return Promise.reject(error)
      })
  }
}

export const tmdbDatasourceFactory = (): TmdbDatasource => {
  if (!(global as any).__tmdbSingleton__) (global as any).__tmdbSingleton__ = new TmdbDatasource()
  return (global as any).__tmdbSingleton__
}

export default tmdbDatasourceFactory()