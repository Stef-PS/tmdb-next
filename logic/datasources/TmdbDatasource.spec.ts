import axios, { AxiosInstance } from 'axios'
import { TmdbDatasource } from './TmdbDatasource'
import { searchMovieResultMock } from './searchMovieResult.mock'

jest.mock('axios', () => ({
  get: jest.fn(() => Promise.resolve({ data: searchMovieResultMock('search', 1, 20, 40) }))
}))
const mockedAxios = axios as jest.Mocked<typeof axios>
const tmdb = new TmdbDatasource()

describe('Tmdb Client', () => {
  describe('saerchMovie', () => {
    it('should perform a movie search with the right parameters', async () => {
      await tmdb.searchMovie('search', 1)
      expect(mockedAxios.get).toHaveBeenCalledWith(
        '/search/movie?api_key=test&query=search&page=1&language=fr-FR',
        { baseUrl: 'https://api.themoviedb.org/3', timeout: 2500, headers: { 'Content-Type': 'application/json', 'Accept': 'application/json' } }
      )
    })

    it('should return a valid movie search', async () => {
      const result = await tmdb.searchMovie('search', 1)
      expect(result).toEqual(searchMovieResultMock('search', 1, 20, 40))
    })

    it('should return the second page results', async () => {
      mockedAxios.get.mockResolvedValueOnce({ data: searchMovieResultMock('search', 2, 20, 40) })
      const { page, results } = await tmdb.searchMovie('search', 2)
      expect(page).toEqual(2)
      expect(results.length).toEqual(20)
      expect(results[0].id).toEqual(20)
    })

    it('should return an empty results on page 3', async () => {
      mockedAxios.get.mockResolvedValueOnce({ data: searchMovieResultMock('search', 3, 20, 40) })
      const { page, results } = await tmdb.searchMovie('search', 3)
      expect(page).toEqual(3)
      expect(results.length).toEqual(0)
    })

    it('should return a limited set of results if total is less than page size', async () => {
      mockedAxios.get.mockResolvedValueOnce({ data: searchMovieResultMock('search', 1, 20, 10) })
      const { page, results } = await tmdb.searchMovie('search', 1)
      expect(page).toEqual(1)
      expect(results.length).toEqual(10)
    })

    it('should reject with the error', async () => {
      const error = {
        status_code: 7,
        status_message: 'Invalid API key: You must be granted a valid key.',
        success: false
      }
      mockedAxios.get.mockRejectedValueOnce(error)
      await expect(tmdb.searchMovie('search', 1)).rejects.toEqual(error)
    })

    it('should return page 1 as default', async () => {
      const { page } = await tmdb.searchMovie('search')
      expect(page).toEqual(1)
    })
  })
})
