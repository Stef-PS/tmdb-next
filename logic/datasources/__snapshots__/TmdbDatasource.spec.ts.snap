// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tmdb Client getMovie should return a valid movie 1`] = `
Object {
  "adult": false,
  "backdropPath": "backdrop_path_1",
  "belongsToCollection": Object {
    "backdropPath": "backdrop_path_1",
    "id": 1,
    "name": "name_1",
    "posterPath": "poster_path_1",
  },
  "budget": 1,
  "genres": Array [
    Object {
      "id": 1,
      "name": "name_1",
    },
  ],
  "homepage": "homepage_1",
  "id": 1,
  "imdbId": "imdb_id_1",
  "originalLanguage": "original_language",
  "originalTitle": "original_title_1",
  "overview": "overview_1",
  "popularity": 1,
  "posterPath": "poster_path_1",
  "productionCompanies": Array [
    Object {
      "id": 1,
      "logoPath": "logo_path_1",
      "name": "name_1",
      "originCountry": "origin_country_1",
    },
  ],
  "productionCountries": Array [
    Object {
      "iso3166_1": "iso_3166_1_1",
      "name": "name_1",
    },
  ],
  "releaseDate": "release_date",
  "revenue": 1,
  "runtime": 1,
  "spokenLanguages": Array [
    Object {
      "englishName": "english_name_1",
      "iso639_1": "iso_639_1_1",
      "name": "name_1",
    },
  ],
  "status": "status",
  "tagline": "tagline_1",
  "title": "title_1",
  "video": false,
  "voteAverage": 1,
  "voteCount": 1,
}
`;
